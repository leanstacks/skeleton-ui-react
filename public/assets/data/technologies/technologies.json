[
  {
    "id": "adbe711c-9698-41f7-bf86-ae3e41ba49eb",
    "name": "React",
    "type": "application",
    "version": "17.0.2",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "React is a structural framework for dynamic web apps.",
    "url": "https://reactjs.org/"
  },
  {
    "id": "0360ff87-0f87-411e-8231-797806583067",
    "name": "Redux",
    "type": "application",
    "version": "4.1.2",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "Redux is a predictable state container for JavaScript apps.",
    "url": "https://redux.js.org/"
  },
  {
    "id": "805356dc-f453-4842-89c8-7d763310c5a3",
    "name": "React Router",
    "type": "application",
    "version": "5.3.0",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "Components are the heart of React's powerful, declarative programming model. React Router is a collection of navigational components that compose declaratively with your application.",
    "url": "https://reacttraining.com/react-router/"
  },
  {
    "id": "4b5b3180-2169-4d80-a22c-1ca4248cb049",
    "name": "Axios",
    "type": "application",
    "version": "0.24.0",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "A Promise-based HTTP client for the browser and node.js.",
    "url": "https://github.com/axios/axios"
  },
  {
    "id": "3b5968c7-6109-428b-bc2b-62652d8c6d8d",
    "name": "Bootstrap",
    "type": "application",
    "version": "5.1.3",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "Bootstrap is an open source toolkit for developing with HTML, CSS, and JS.",
    "url": "https://getbootstrap.com/"
  },
  {
    "id": "7d09624e-0256-4fe5-9b9f-bf76683536a1",
    "name": "Font Awesome",
    "type": "application",
    "version": "5.15.4",
    "license": "Multiple",
    "licenseUrl": "https://fontawesome.com/license",
    "description": "Font Awesome is an icon toolkit.",
    "url": "https://fontawesome.com/"
  },
  {
    "id": "30130c49-cacf-474e-928c-c4a0190914ef",
    "name": "Animate.css",
    "type": "application",
    "version": "4.1.1",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "Animate.css is a library of ready-to-use, cross-browser animations for use in your web projects. Great for emphasis, home pages, sliders, and attention-guiding hints.",
    "url": "https://animate.style/"
  },
  {
    "id": "8f2d877d-d17a-4b3c-bbc3-690f460a57e0",
    "name": "Lodash",
    "type": "application",
    "version": "4.17.21",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "A modern JavaScript utility library delivering modularity, performance & extras.",
    "url": "https://lodash.com/"
  },
  {
    "id": "aa636860-0fa5-4093-9592-1cbb6b2f5c3b",
    "name": "Luxon",
    "type": "application",
    "version": "2.1.1",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "A powerful, modern, and friendly wrapper for JavaScript dates and times.",
    "url": "https://moment.github.io/luxon"
  },
  {
    "id": "d142466a-e461-474b-9aac-3490a37906d8",
    "name": "Webpack",
    "type": "build",
    "version": "4.46.0",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "Webpack is a static module bundler for modern web applications.",
    "url": "https://webpack.js.org/"
  },
  {
    "id": "7671e7af-7dc6-4d22-ba65-6a21778d2beb",
    "name": "Babel",
    "type": "build",
    "version": "7.16.0",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "Babel is a JavaScript transpiler.",
    "url": "https://babeljs.io/"
  },
  {
    "id": "126d771c-b606-4c10-866b-4385e44b15d3",
    "name": "Node.js",
    "type": "build",
    "version": "14.16.0",
    "description": "Node.js is a JavaScript runtime.",
    "url": "https://nodejs.org/"
  },
  {
    "id": "6cab5852-7750-4af3-b5dc-59220f2ca03e",
    "name": "Yarn",
    "type": "build",
    "version": "1.22.17",
    "license": "BSD 2-Clause",
    "licenseUrl": "https://spdx.org/licenses/BSD-2-Clause.html",
    "description": "Fast, reliable, and secure dependency management. Yarn is an alternative to the npm client.",
    "url": "https://yarnpkg.com/"
  },
  {
    "id": "8218cb49-7830-4d88-8df5-4c1ad4d4df4f",
    "name": "Jest",
    "type": "test",
    "version": "27.3.1",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "Jest is used by Facebook to test all JavaScript code including React applications.",
    "url": "http://jestjs.io/"
  },
  {
    "id": "f62dd51f-ceae-4051-bbf1-931161ee54fc",
    "name": "Enzyme",
    "type": "test",
    "version": "3.11.0",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "Enzyme is a JavaScript Testing utility for React that makes it easier to assert, manipulate, and traverse your React Components' output.",
    "url": "http://airbnb.io/enzyme/"
  },
  {
    "id": "878a061d-2605-4839-ba44-8b2eced0e982",
    "name": "Redux Mock Store",
    "type": "test",
    "version": "1.5.4",
    "license": "MIT",
    "licenseUrl": "https://spdx.org/licenses/MIT.html",
    "description": "A mock store for testing Redux async action creators and middleware.",
    "url": "https://www.npmjs.com/package/redux-mock-store"
  }
]