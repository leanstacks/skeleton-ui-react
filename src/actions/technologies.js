import axios from 'axios';

// SET_TECHNOLOGIES
export const setTechnologies = (technologies) => ({
  type: 'SET_TECHNOLOGIES',
  technologies
});

// START_SET_TECHNOLOGIES
export const startSetTechnologies = () => {
  return (dispatch) => {
    const config = {
      url: `/assets/data/technologies/technologies.json`,
      method: 'get'
    };
    return axios.request(config)
      .then((response) => {
        dispatch(setTechnologies(response.data));
        localStorage.setItem('technologies', JSON.stringify(response.data));
        localStorage.setItem('technologies_lu', new Date().toISOString());
      }).catch((err) => {
        console.error('API error. ', err);
      });
  };
};
