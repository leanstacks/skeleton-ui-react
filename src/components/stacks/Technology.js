import React from 'react';

const Technology = (props) => (
  <div className="row">
    <div className="col">
      <h4>{props.technology.name}</h4>
      {props.technology.version &&
        <div><span className="fw-bold">Version:</span>&nbsp;{props.technology.version}</div>
      }
      {props.technology.license &&
        <div>
          <span className="fw-bold">License:</span>&nbsp;{props.technology.license}
          {props.technology.licenseUrl &&
            <a className="text-decoration-none" href={props.technology.licenseUrl} title={props.technology.license} target="_blank">&nbsp;<i className="fas fa-external-link-alt"></i></a>
          }
        </div>
      }
      <div className="lead mt-3">{props.technology.description}</div>
      {props.technology.url &&
        <div className="text-right mt-3">
          <a className="text-decoration-none" href={props.technology.url} title={props.technology.name} target="_blank">Go to website&nbsp;<i className="fas fa-external-link-alt"></i></a>
        </div>
      }
    </div>
  </div>
);

export default Technology;