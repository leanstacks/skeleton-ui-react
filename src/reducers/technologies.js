const defaultState = [];

export default (state = defaultState, action) => {
  switch (action.type) {
    case 'SET_TECHNOLOGIES':
      return action.technologies;
    default:
      return state;
  };
};
