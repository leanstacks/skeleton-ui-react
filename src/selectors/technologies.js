import filter from 'lodash/filter';
import find from 'lodash/find';

export const getTechnologiesByType = (technologies = [], type) => {
  return filter(technologies, { type });
};

export const getTechnology = (technologies = [], technologyId) => {
  return find(technologies, { id: technologyId });
};
