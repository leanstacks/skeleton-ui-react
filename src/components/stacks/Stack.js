import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

import { getTechnologiesByType } from '../../selectors/technologies';

export const Stack = (props) => {
  const { name, technologies = [], type } = props;
  return (
    <div>
      <h3>{name} Stack</h3>
      <ul className="list-unstyled">
        {technologies.map(technology => <li key={technology.id}>{technology.name}</li>)}
      </ul>
      <Link className="text-decoration-none" to={`/stacks/${type}`}>Learn More&nbsp;<i className="fas fa-caret-right"></i></Link>
    </div>
  );
}

const mapStateToProps = (state, props) => ({
  technologies: getTechnologiesByType(state.technologies, props.type)
});

export default connect(mapStateToProps)(Stack);