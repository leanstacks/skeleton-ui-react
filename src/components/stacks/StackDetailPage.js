import React from 'react';
import { connect } from 'react-redux';
import { Link, withRouter } from 'react-router-dom';
import capitalize from 'lodash/capitalize';

import { getTechnologiesByType, getTechnology } from '../../selectors/technologies';
import Technology from './Technology';

export const StackDetailPage = (props) => (
  <div className="page">
    <div className="container">
      <div className="row mt-5 mb-3">
        <div className="col-12">
          <h2 className="display-4">
            {capitalize(props.stackType)} Stack
          </h2>
          <div className="lead">Select any technology to view additional detail.</div>
        </div>
      </div>

      <div className="row mb-3">
        <div className="col-12 col-md-6 mb-3 mb-md-0">
          <h3>Technologies</h3>
          <ul className="list-unstyled mb-4">
            {
              props.technologies.map((technology) => (
                <li key={technology.id}>
                  <Link className="text-decoration-none" to={`/stacks/${props.stackType}/technologies/${technology.id}`}>{technology.name}</Link>
                </li>
              ))
            }
          </ul>
          <div>
            <Link className="text-decoration-none" to="/stacks"><i className="fas fa-caret-left"></i>&nbsp;Back</Link>
          </div>
        </div>

        {props.technology &&
          <div className="col-12 col-md-6 mb-3 mb-md-0">
            <div className="d-none d-md-block ps-2 py-3 border-2 border-start border-secondary h-100">
              <Technology technology={props.technology} />
            </div>
            <div className="d-md-none mt-2 pt-2 border-2 border-top border-secondary">
              <Technology technology={props.technology} />
            </div>
          </div>
        }
      </div>
    </div>
  </div>
);

const mapStateToProps = (state, props) => ({
  stackType: props.match.params.stackType,
  technologies: getTechnologiesByType(state.technologies, props.match.params.stackType),
  technology: getTechnology(state.technologies, props.match.params.technologyId)
});

export default withRouter(connect(mapStateToProps)(StackDetailPage));