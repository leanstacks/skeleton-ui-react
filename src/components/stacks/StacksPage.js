import React from 'react';

import Stack from './Stack';

const StacksPage = (props) => (
  <div className="page">
    <div className="container">
      <div className="row mt-5 mb-3">
        <div className="col-12">
          <h2 className="display-4">Technology Stacks</h2>
        </div>
      </div>
      <div className="row mb-3">
        <div className="col-12 col-md-4 mb-3 mb-md-0">
          <Stack name="Application" type="application" />
        </div>
        <div className="col-12 col-md-4 mb-3 mb-md-0">
          <Stack name="Test" type="test" />
        </div>
        <div className="col-12 col-md-4 mb-3 mb-md-0">
          <Stack name="Build" type="build" />
        </div>
      </div>
    </div>
  </div>
);


export default StacksPage;