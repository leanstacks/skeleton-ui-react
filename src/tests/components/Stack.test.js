import React from 'react';
import { shallow } from 'enzyme';

import { Stack } from '../../components/stacks/Stack';
import technologies from '../fixtures/technologies';

test('should render Stack correctly', () => {
  const wrapper = shallow(<Stack technologies={technologies} type="application" />);
  expect(wrapper).toMatchSnapshot();
});