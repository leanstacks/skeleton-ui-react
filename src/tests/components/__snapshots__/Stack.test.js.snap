// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Stack correctly 1`] = `
<div>
  <h3>
     Stack
  </h3>
  <ul
    className="list-unstyled"
  >
    <li
      key="adbe711c-9698-41f7-bf86-ae3e41ba49eb"
    >
      React
    </li>
    <li
      key="0360ff87-0f87-411e-8231-797806583067"
    >
      Redux
    </li>
    <li
      key="805356dc-f453-4842-89c8-7d763310c5a3"
    >
      React Router
    </li>
    <li
      key="4b5b3180-2169-4d80-a22c-1ca4248cb049"
    >
      Axios
    </li>
    <li
      key="3b5968c7-6109-428b-bc2b-62652d8c6d8d"
    >
      Bootstrap
    </li>
    <li
      key="7d09624e-0256-4fe5-9b9f-bf76683536a1"
    >
      Font Awesome
    </li>
    <li
      key="8f2d877d-d17a-4b3c-bbc3-690f460a57e0"
    >
      Lodash
    </li>
    <li
      key="aa636860-0fa5-4093-9592-1cbb6b2f5c3b"
    >
      Moment
    </li>
    <li
      key="e7b38c23-e22d-4cc3-9626-dffce5c5bd2a"
    >
      Numeral
    </li>
    <li
      key="d142466a-e461-474b-9aac-3490a37906d8"
    >
      Webpack
    </li>
    <li
      key="7671e7af-7dc6-4d22-ba65-6a21778d2beb"
    >
      Babel
    </li>
    <li
      key="126d771c-b606-4c10-866b-4385e44b15d3"
    >
      Node.js
    </li>
    <li
      key="079dd1b7-f5b1-4c32-a0af-c7ea65e94976"
    >
      NPM
    </li>
    <li
      key="6cab5852-7750-4af3-b5dc-59220f2ca03e"
    >
      Yarn
    </li>
    <li
      key="8218cb49-7830-4d88-8df5-4c1ad4d4df4f"
    >
      Jest
    </li>
    <li
      key="f62dd51f-ceae-4051-bbf1-931161ee54fc"
    >
      Enzyme
    </li>
    <li
      key="878a061d-2605-4839-ba44-8b2eced0e982"
    >
      Redux Mock Store
    </li>
  </ul>
  <Link
    className="text-decoration-none"
    to="/stacks/application"
  >
    Learn MoreÂ 
    <i
      className="fas fa-caret-right"
    />
  </Link>
</div>
`;
